docs: rewrite state persistence (#1237)
apify/apify-docs: sources/platform/actors/development/builds_and_runs/state_persistence.md

---
title: State persistence
description: Learn how to maintain an Actor's state to prevent data loss during unexpected restarts. Includes code examples for handling server migrations.
slug: /actors/development/builds-and-runs/state-persistence
---

# State persistence

**Learn how to maintain an Actor's state to prevent data loss during unexpected restarts. Includes code examples for handling server migrations.**

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

---

> Long-running [Actor](../../index.mdx) jobs may need to migrate between servers. Without state persistence, your job's progress, is lost during migration, causing it to restart from the beginning on the new server. This can be costly and time-consuming.

Reviewer: ```suggestion
Long-running [Actor](../../index.mdx) jobs may need to migrate between servers. Without state persistence, your job's progress is lost during migration, causing it to restart from the beginning on the new server. This can be costly and time-consuming.
```

To prevent data loss, long-running Actors should:

- Periodically save (persist) their state.
- Listem for [migration events](/sdk/js/api/apify/class/PlatformEventManager)
- Check for persisted state when starting, allowing them to resume from where they left off.

For short-running Actors, the risk of restarts and the cost of repeated runs are low, so you can typically ignore state persistence.

> ## Undersanding migrations

Reviewer: ```suggestion
## Understanding migrations
```

> A migration occurs when a process running on one srever must stop and move to another. During this process:

Reviewer: ```suggestion
A migration occurs when a process running on one server must stop and move to another. During this process:
```

- All in-progress processes on the current server are stopped

> - Unless you've saved your state, the Actor run will restart on the new server

Reviewer: The Actor will restart even if you've saved your state, the only difference is how
```suggestion
- Unless you've saved your state, the Actor run will restart on the new server with an empty internal state
```

- You only have a few seconds to save your work when a migration event occurs

### Causes of migration

Migrations can happen for several reasons:

- Server workload optimization
- Server crashes (rare)
- New feature releases and bug fixes

### Frequency of migrations

Migrations don't follow a specific schedule. They can occur at any time due to the events mentioned above.

## Why state is lost during migration

> By default, an Actor keeps its output and state in the server's memory. During a server switch, the run loses access to the previous server's memory. Even if data were saved on the server's disk, access to that would also be lost.

Reviewer: I tried to clarify this a bit, but feel free to rewrite:
```suggestion
By default, an Actor keeps its state in the server's memory. During a server switch, the run loses access to the previous server's memory. Even if data were saved on the server's disk, access to that would also be lost. Note that the Actor run's default dataset, key-value store and request queue are preserved across migrations, by state we mean the contents of runtime variables in the Actor's code.
```

## Implementing state persistence

The [Apify SDKs](/sdk) handle state persistence automatically.

This is done  using the `Actor.on()` method and the `migrating` event.

- The `migrating` event is triggered just before a migration occurs, allowing you to save your state.

> - To retrieve previously saved state, you can use the [`Actor.getValue`](/sdk/js/reference/class/Actor#getValue)/[`Actor.get_value()`](/sdk/python/reference/class/Actor#get_value) methods.

Reviewer: I unified this to both have no parentheses
```suggestion
- To retrieve previously saved state, you can use the [`Actor.getValue`](/sdk/js/reference/class/Actor#getValue)/[`Actor.get_value`](/sdk/python/reference/class/Actor#get_value) methods.
```

### Code examples

> To manually persis state, use the `Actor.on` method in the Apify SDK:

Reviewer: ```suggestion
To manually persist state, use the `Actor.on` method in the Apify SDK:
```

- Server crashes (rare)
- New feature releases and bug fixes

### Frequency of migrations

Migrations don't follow a specific schedule. They can occur at any time due to the events mentioned above.

## Why state is lost during migration

By default, an Actor keeps its output and state in the server's memory. During a server switch, the run loses access to the previous server's memory. Even if data were saved on the server's disk, access to that would also be lost.

## Implementing state persistence

The [Apify SDKs](/sdk) handle state persistence automatically.

In JavaScript, this is done using the `migrating` and `persistState` events in the [PlatformEventManager](/sdk/js/api/apify/class/PlatformEventManager).

- The `persistState` event prompts SDK components to save their state at regular intervals
- The `migrating` event is triggered just before a migration occurs.

In Python, state persistence is handled using the `Actor.on()` method and the migrating event, similar to JavaScript. The Apify SDK for Python provides mechanisms to save and retrieve state data.

- The `migrating` event is triggered just before a migration occurs, allowing you to save your state.

> - To retrieve previously saved state, you can use the `Actor.get_value()` method.

Reviewer 2: I believe you could link the `PlatformEventManager` API reference of both JS and Python implementations.

You can also mention, that the event manager is encapsulated with Actor class, and when you want to interact with that you should use `Actor.on` and `Actor.off`. And it will use the underlying event manager instance in Actor.

Or you do not have to mention `EventManagers` at all and go with just the shortcuts.
